### ISSUE TITLE
Orphan .json and .xml files left in /tbldefs cause Duplicate Table error on Build From Source

### ISSUE THREAD

**Issue by jhgarrison** [joyfullservice/msaccess-vcs-integration#296]
Version 3.4.18

I've managed to avoid using the Build From Source option until now, but I decided it was time to make sure Build From Source could actually rebuild my database :-).  I ran into one problem, with a workaround.

I've been working on modernizing a legacy database, during which I reorganized the tables.
In some cases tables that were linked were moved to the front-end, and in others tables were pushed to a back-end.

When I went to build from source for the first time I got a "Duplicate Table" error when the plugin first created the local table and then tried to create a link with the same name.  Looking in the /tbldefs directory I see a number of tables with THREE files, .json, .sql and .xml.  These are all tables whose location changed at some point.  The orphan file was causing the build to attempt to create the table twice.

I worked around this by manually deleting the offending .json (for originally linked tables which are now local) or .xml file (for originally local tables that are now links).  After that the Build From Source completed successfully.

**Comment by jhgarrison** [comment]
In case it's useful for anyone, here's a bash (Cygwin) one-liner to list the tables with orphan files:

`for i in *.sql; do j=${i%.sql}; [[ -e $j.xml && -e $j.json ]] && echo $i; done`

This just lists the table names (the .sql files), you have to know which orphan to delete based on where the table actually resides.

**Comment by hecon5** [comment]
Which version of Access are you using again? I am pretty sure you said...but my memory's a bit foggy :)

I also have this problem, but in my experience, it's almost always self-inflicted. Can you try these and see if the tool is still exporting orphans?

1. Do a `Full Export` from time to time. I occasionally find the difference engine (hash) doesn't catch all the changes. Linked tables appear to be the most prone to this in my experience. Tables are an older component and I don't know if they've been integrated fully into the hash compare, and from time to time are not exported, even if they should be. @joyfullservice, am I being accurate here?
2. Prior to a `Full Export`, go through git and commit all changes you want, discard everything else. Effectively, get to a clean slate. I sometimes discover there's a file I never committed and git helpfully puts back to the old state when I switch branches, or leaves there because it's not in the repository yet (an extremely common cause of my doing this is when I am working on a new form and the initial stages of this I'll just bang a whole ton of changes at once while I'm prototyping). This self-inflicted wound will then come back to bite me when it's left behind and I don't remove it prior to doing a build. 
3. Tables are my nemesis for item 2: we use a LOT of linked tables for forms, and their connection strings are ... temperamental at best due to the way Access links them. This makes noise on the export, and I thusly ignore table changes simply because at first glance they appear to be noise. Classic case of ignoring the kid who cried wolf.
4. Switching branches but leaving uncommitted changes (I don't do a clean switch). Try reverting / removing all non-committed items. This effectively leaves orphans around if I don't then go through and commit/discard anything I don't want to actually be added to the repository.

**Comment by hecon5** [comment]
To follow up and answer the rest of your question on the three file types: 

|File Type |Used For |Details & Description
|-----|:-:|:---
|`JSON` |Linked Tables Only |Contains additional table attributes (should be all, but I'm sure there's probably an edge case that's not been caught yet), such as connection string, the Access table name, the linked table name (on the server), a list of attributes (it's a `long` type that's used in `bitwise` fashion), and a few other bits to cue MSAccessVCS how to import the table.
|`XML` |Local Tables Only |Table format and extra information (such as row display height, column display width, required/not, column formatting information (Plain Text/RTF, Date Time formatting, etc.) and so on. It is the most 'reliable' and full export of the local table as it (usually) contains all the meta data about the table. `XML` ensures pretty much all the format and data will be the same when you rebuild. If you export the table data, too (it's stored in a separate folder), then `XML` is the way to go (usually). You can also export Table data via `TDF` (Tab Delimited Format) if you want something more human-readable. `XML` by its nature is not as row or character dependent. Access also isn't careful at exporting the same components in the same way every time, so the export is mildly prone to code noise during export (AKA: `git` will think there's changes, when there aren't).
|`SQL` |Linked and Local Tables|If you were to generate a table via SQL, this would be the SQL you'd use. Because Access SQL (SQL Jet) doesn't care about formatting/display width/display height, if you use this, the table won't "look" the same when you open it in Access.

**Comment by joyfullservice** [comment]
@hecon5 - A couple comments to add...

> 1. Do a `Full Export` from time to time. I occasionally find the difference engine (hash) doesn't catch all the changes. Linked tables appear to be the most prone to this in my experience. Tables are an older component and I don't know if they've been integrated fully into the hash compare, and from time to time are not exported, even if they should be. @joyfullservice, am I being accurate here?
> 

The hashing and change tracking is reworked pretty significantly in version 4.0 (currently in development) so this may be improved or resolved in the next major release. If not, we can open an issue and work through it.

> 
> 3. Tables are my nemesis for item 2: we use a LOT of linked tables for forms, and their connection strings are ... temperamental at best due to the way Access links them. This makes noise on the export, and I thusly ignore table changes simply because at first glance they appear to be noise. Classic case of ignoring the kid who cried wolf.
> 

In some of my projects I have a startup routine that verifies the table links to make sure they all match the desired connection string, updating as necessary. This has worked well for me in reducing unwanted noise with changes in table connection strings.

```vba
'---------------------------------------------------------------------------------------
' Procedure : UpdateConnectionStrings
' Author    : Adam Waller
' Date      : 9/30/2019
' Purpose   : Make sure we are linked to the correct set of SQL tables.
'---------------------------------------------------------------------------------------
'
Public Sub UpdateConnectionStrings()

    Dim tdf As TableDef
    Dim qry As QueryDef
    Dim strConnect As String
    Dim strNew As String
    Dim strUtilityDB As String
    Dim strCustomDbConnection As String
    Dim strSEDbConnection As String
    Dim strFile As String
    Dim strDataPath As String
    
    ' Look up database names
    strUtilityDB = SE_API.Utility.AdoConn.DefaultDatabase
    
    ' Get connection string for current environment
    strCustomDbConnection = GetADOConnectionString(SE_API.Utility.AdoConn)
    strSEDbConnection = SE_API.ADO.ConnectionString
    
    ' Go through tables
    For Each tdf In CodeDb.TableDefs
        If Left(tdf.Name, 1) <> "~" And Left(tdf.Name, 1) <> "z" Then
            strConnect = tdf.Connect
            If strConnect <> "" Then
                If Left(strConnect, 5) = "ODBC;" Then
                    If InStr(1, strConnect, "=" & strUtilityDB & ";", vbTextCompare) < 1 Then
                        If strConnect <> strCustomDbConnection Then
                            Debug.Print tdf.Name & " - " & tdf.Connect
                            tdf.Connect = strCustomDbConnection
                            tdf.RefreshLink
                        End If
                    End If
                ElseIf Left(strConnect, 10) = ";DATABASE=" Then
                    ' Could be a data table
                    strFile = GetFileNameFromPath(strConnect)
                    If strFile Like "* Data.accdb" Then
                        ' Link to data table
                        If strDataPath = vbNullString Then strDataPath = SE_API.Utility.GetValue("Data Folder", , "GLOBAL")
                        strNew = ";DATABASE=" & strDataPath & strFile
                        If strConnect <> strNew Then
                            tdf.Connect = strNew
                            tdf.RefreshLink
                        End If
                    End If
                End If
            End If
        End If
    Next tdf
    
    ' Go through queries
    For Each qry In CodeDb.QueryDefs
        If Left(qry.Name, 1) <> "~" And Left(qry.Name, 1) <> "z" Then
            strConnect = qry.Connect
            If strConnect <> "" Then
                If Left(strConnect, 5) = "ODBC;" Then
                    If IsLinkedToNonSETable(strConnect) Then
                        strNew = strCustomDbConnection
                    Else
                        strNew = strSEDbConnection
                    End If
                    ' Update connmection string if different.
                    If strConnect <> strNew Then
                        Debug.Print qry.Name & " - " & qry.Connect
                        qry.Connect = strNew
                    End If
                End If
            End If
        End If
    Next qry
    
End Sub
```

**Comment by hecon5** [comment]
Oh Slick! Thank you; I'll give this a try and see if it helps. Since we have a lot of linked tables, I'm concerned it will take eons to run, but I'll see and report back on this!

**Comment by joyfullservice** [comment]
@jhgarrison - Thanks for pointing out this issue! As you discovered, the output files are different types, depending on whether the table is a linked or local table. When building from source, the database won't know which format it was supposed to use since both source files are technically valid.

The files also don't get cleared out through the routine that removes orphaned files because the underlying table object still exists with the same name in the database. The approach I am thinking here is to have the add-in remove the alternate source file during the export. If you export a linked file, it should remove the XML file as orphaned. If you export a local table, it should remove the linked file as orphaned.

**Comment by joyfullservice** [comment]
> Oh Slick! Thank you; I'll give this a try and see if it helps. Since we have a lot of linked tables, I'm concerned it will take eons to run, but I'll see and report back on this!

The nice thing about this approach is that it only relinks the table _if the connection string is different_. It will take a while the first time you run it, since it may need to fix a number of connection strings, but after that, it should be pretty fast since it is just scanning and not actually changing or relinking the tables.

**Comment by hecon5** [comment]
I see that now; I've been simply ignoring it and reverting them upon rebuild, but this would work considerably better. 
We have an internal table with the "correct" connection strings so we can switch between test and production environments, but for whatever reason, Access rearranges the items in the connection string after doing this.

**Comment by jhgarrison** [comment]
> Which version of Access are you using again? I am pretty sure you said...but my memory's a bit foggy :)

The latest and greatest Office 365

> 
> I also have this problem, but in my experience, it's almost always self-inflicted. Can you try these and see if the tool is still exporting orphans?
> 
>     1. Do a `Full Export` from time to time.

I always do a full export at every release point.  This error occurred immediately after a full export.

> 
>     2. Prior to a `Full Export`, go through git and commit all changes you want, discard everything else. 

Ditto, I always do this

>     3. Tables are my nemesis for item 2: we use a LOT of linked tables for forms ...

I also wrote a re-link procedure that gets run at startup every time.  Eliminates the need to manually manage links.

>     4. Switching branches but leaving uncommitted changes (I don't do a clean switch).

Nope, I never do that.  

Here's the full list of steps I did.

1. In the original directory, do a full export, checkin all changes, push everything to github.
2. Create a new directory and clone the repo from github
3. Copy all the backend .accdb files to the new directory
4. In the new directory, create an empty front-end .accdb and then Build from Source

This is where I encountered the errors.

Whatever the cause, I'm not sure it's worth fixing given the work on 4.0 and easy workaround.  The files are clearly forgotten orphans because if I do a full export from the newly built DB copy it does not create any orphans.
